### ISSUE TITLE
set-destination SCHEME doesn't work with valid dynamic values

### ISSUE THREAD

**Issue by mal** [apache/trafficserver#4143]
I'm trying to alter the scheme of a request based on a custom header, thus:

```
cond %{READ_REQUEST_HDR_HOOK}
cond %{HEADER:X-ATS-Scheme} /http/
set-destination SCHEME %{CLIENT-HEADER:X-ATS-Scheme}
```

Unfortunately making a request:
```
$ http_proxy=http://traffic:8888 curl -i google.com -H 'X-ATS-Scheme: http'
```

Results in a `HTTP/1.1 400 Unsupported URL Scheme`. However changing the final line to be:

```
set-destination QUERY %{HEADER:X-ATS-Scheme}
```

Happily rewrites the request to `http://google.com/?http`. And hardcoding the scheme in the `set-destination`:

```
set-destination SCHEME http
```

Works as I'd hoped the dynamic version would. It seems like this should be possible but is perhaps a bug in that it doesn't recognise when the dynamic value is a valid scheme?

---

Tested on:
ATS 7.0.0 as provided by the default package repo for Debian Stretch
ATS 7.1.2 as provided by the default package repo for Ubuntu 18.04

**Comment by mlibbey** [comment]
Know its not exactly what you are after, but what about 
cond %{READ_REQUEST_HDR_HOOK} [AND]
cond %{HEADER:X-ATS-Scheme}="http"
set-destination SCHEME http

cond %{READ_REQUEST_HDR_HOOK} [AND]
cond %{HEADER:X-ATS-Scheme}="https"
set-destination SCHEME https

**Comment by zwoop** [comment]
Yeh, this is a bug, sorry for the delay ,but I'll have a fix for 9.0.0 soon.

**Comment by zwoop** [comment]
```
diff --git a/plugins/header_rewrite/operators.cc b/plugins/header_rewrite/operators.cc
index c2050116b..1d7ccf320 100644
--- a/plugins/header_rewrite/operators.cc
+++ b/plugins/header_rewrite/operators.cc
@@ -239,26 +239,28 @@ OperatorSetDestination::exec(const Resources &res) const
       }
       break;
     case URL_QUAL_URL:
-      if (_value.empty()) {
+      _value.append_value(value, res);
+      if (value.empty()) {
         TSDebug(PLUGIN_NAME, "Would set destination URL to an empty value, skipping");
       } else {
-        const char *start = _value.get_value().c_str();
-        const char *end   = _value.get_value().size() + start;
+        const char *start = value.c_str();
+        const char *end   = start + value.size();
         TSMLoc new_url_loc;
         if (TSUrlCreate(bufp, &new_url_loc) == TS_SUCCESS && TSUrlParse(bufp, new_url_loc, &start, end) == TS_PARSE_DONE &&
             TSHttpHdrUrlSet(bufp, res.hdr_loc, new_url_loc) == TS_SUCCESS) {
-          TSDebug(PLUGIN_NAME, "Set destination URL to %s", _value.get_value().c_str());
+          TSDebug(PLUGIN_NAME, "Set destination URL to %s", value.c_str());
         } else {
-          TSDebug(PLUGIN_NAME, "Failed to set URL %s", _value.get_value().c_str());
+          TSDebug(PLUGIN_NAME, "Failed to set URL %s", value.c_str());
         }
       }
       break;
     case URL_QUAL_SCHEME:
-      if (_value.empty()) {
+      _value.append_value(value, res);
+      if (value.empty()) {
         TSDebug(PLUGIN_NAME, "Would set destination SCHEME to an empty value, skipping");
       } else {
-        TSUrlSchemeSet(bufp, url_m_loc, _value.get_value().c_str(), _value.get_value().length());
-        TSDebug(PLUGIN_NAME, "OperatorSetDestination::exec() invoked with SCHEME: %s", _value.get_value().c_str());
+        TSUrlSchemeSet(bufp, url_m_loc, value.c_str(), value.length());
+        TSDebug(PLUGIN_NAME, "OperatorSetDestination::exec() invoked with SCHEME: %s", value.c_str());
       }
       break;
     default:
```
